src/core/namespace.c

```diff
@@ -816,18 +816,24 @@ static int make_read_only(MountEntry *m, char **blacklist) {
         return r;
 }
 
-static bool namespace_info_mount_apivfs(const NameSpaceInfo *ns_info) {
+static bool namespace_info_mount_apivfs(const char *root_directory, const NameSpaceInfo *ns_info) {
         assert(ns_info);
 
-        /* ProtectControlGroups= and ProtectKernelTunables= imply MountAPIVFS=, since to protect the API VFS mounts,
-         * they need to be around in the first place... */
+        /*
+         * ProtectControlGroups= and ProtectKernelTunables= imply MountAPIVFS=,
+         * since to protect the API VFS mounts, they need to be around in the
+         * first place... and RootDirectory= or RootImage= need to be set.
```

Comment:
@keszybz I tend to agree with you here, but I thought that the discussion about MountAPIVFS= was closed since it was merged! hence my patches. I would also prefer to not set PrivateDevices=yes explicitly but make it the default if we are in the ProtectSomething=true mode.

> Dunno, rather then pile fixes on top, I'd rather revert the change that MountAPIVFS=yes is implied by Protect*=yes.

I'm fine with this too :-)  but it's not related to this patch.

> When RootImage=/RootDirectory= are not set, all the API filesystems must be already mounted, so there's nothing to do. PID1 ensures that this is true in bootup.

Yes but the current code doesn't check RootImage=/RootDirectory= and tries to do something even under PID 1 or in the init mount namespace of PID1 and it will fail. /proc, /sys and /dev are already mount points. So this patch 2/2 just makes sure to avoid the unnecessary steps and noise for this. This also allows it to fix the case of RootImage=/RootDirectory which are not in the other inferior mount namespaces.

> So let's just say that MountAPIVFS= applies only to the case when RootImage=/RootDirectory= are set, and must be explicitly set if wanted.

That's what this patch is doing :-) 

I think you don't like the fact that MountAPIVFS= is implied in case of ProtectSomething=true, I understand. This is not related to this patch. Patch 2/2 makes sure that MountAPIVFS= is set only if RootDirectory/RootImage= , so it's not related at all in this patch.

I dropped that patch that also made PrivateDevices=yes if MountAPIVFS= is implied, to reduce any unexpected behaviour or user surprises. That can be addressed in a separate patch or discussion of course. **Discuss MountAPIVFS= implicit behaviour**